state/azure/tablestorage/tablestorage.go

```diff
@@ -195,20 +193,25 @@ func (r *StateStore) writeRow(req *state.SetRequest) error {
 	}
 	entity.OdataEtag = etag
 
-	// InsertOrReplace does not support ETag concurrency, therefore we will try to use Update method first
-	// as it's more frequent, and then Insert
+	// InsertOrReplace does not support ETag concurrency, therefore we will use Insert to check for key existence
+	// and then use Update to update the key if it exists with the specified ETag
 
-	err := entity.Update(false, nil)
+	err := entity.Insert(storage.FullMetadata, nil)
 	if err != nil {
-		if isNotFoundError(err) {
-			// When entity is not found (set state first time) create it
-			entity.OdataEtag = ""
-
-			return entity.Insert(storage.FullMetadata, nil)
+		if etag == "" {
+			return state.NewETagError(state.ETagMismatch, err)
```

Comment:
Doesn't this depend on whether the concurrency state is FirstWrite? (i.e. it should be a valid request to upsert with default last write wins semantics and no ETag specified, the function should have called Update instead).

It doesn't look like the component checks for req.Options.Concurrency anywhere in relation to its ETag handling, which seems wrong.

Reply:
Thanks for reminding me of that table. I now covered the combination of etag set / not set and concurrency options.